### PR TITLE
[CUDAGraph] add autocudagraph

### PR DESCRIPTION
**Pull request by DrRyanHuang**
### Description
Add `autocudagraph` decorator for automated CUDAGraph acceleration of forward and backward passes.

### 🌟 Key Features & Implementation

**`src/paddlefleet/cudagraph.py`：核心基础设施**
* **核心上下文 (`CUDAGraphContext`)**：基于 `dataclass` 统一管理静态 Tensor Buffer、Fwd/Bwd 句柄与 PyLayer Runner。
* **泛型 IO 解析 (`get_tensors` / `set_tensors`)**：支持递归遍历嵌套结构（Tensor / list / tuple / dict）提取与写回。
* **全自动捕获流 (`autocudagraph`)**：支持 warmup -> capture -> replay 完整生命周期。
* **内存优化**：前反向分别捕获为独立 CUDAGraph，共享同一 unique memory pool。
* **动态路由 & Fallback**：引入 `dispatch_key_fn`，超出 `max_graphs` 上限时自动无缝回退 Eager 模式。
* **梯度累积保障**：自动快照与恢复 `nn.Layer` 参数梯度，防止 dummy backward 污染真实的跨步累积梯度。
* **反向传播桥接**：基于 `paddle.autograd.PyLayer` 正确路由输入张量梯度与权重参数梯度。

### 🧪 Test Coverage 
**`tests/single_card_tests/test_autocudagraph.py`**

| Test Class | 覆盖点 |
|:---|:---|
| `TestPureFunctions` | 纯函数、多输入、`stop_gradient` 混合场景 |
| `TestOOPAndSubmodules` | OOP 调用、权重指针漂移校验、多实例隔离 |
| `TestConfigurations` | `dispatch_key_fn` 动态图分派、`max_graphs` Fallback 机制 |
| `TestComplexStructures` | 嵌套 dict/list IO、`no_grad` 上下文对齐 |
| `TestFatModel` | 多层 FFN 模拟、参数梯度端到端对齐 |
| `TestAdvancedMechanics` | 跨步梯度累积正确性、AMP 混合精度支持 |
| `TestEdgeMemoryOps` | Tensor slice/stride 等复杂内存视角操作 |
| `TestGraphLimitationsAndFeatures` | Dropout 随机性保持（兼容 Philox RNG） |
| `TestDispatchNumber` / `TestNoGrad` | `max_graphs` 缓存边界限制、`no_grad` 下的对齐 |

### 📝 Test Plan
- [x] **单测全过**：单卡 GPU 执行 `test_autocudagraph.py`，所有 Case 通过。
- [x] **数值对齐**：不同 `warmup_steps` 下，CG 与 Eager 模式前反向 100% 绝对一致 (`rtol=0`, `atol=0`)。
- [x] **Fallback 鲁棒性**：输入组合超 `max_graphs` 限制时，正常执行不崩溃，平滑回退至 Eager 模式。
- [x] **梯度安全**：多步梯度累积下，累积梯度免受 Capture 阶段内部 Backward 的污染与清零。

### CODE REVIEW

File: tests/single_card_tests/test_autocudagraph.py [hunk 1]
```diff
@@ -0,0 +1,895 @@
...
+        for step in range(total_steps):
+            x = dummy_inputs[step].clone().detach()
+            y = dummy_targets[step]
+
+            out = model_eager(x)
+            loss = F.cross_entropy(out, y)
+            loss.backward()
+
+            losses_eager.append(loss.item())
+            grads_eager.append(model_eager.head2.weight.grad.clone().cpu())
```
Comment (DrRyanHuang) [inline tests/single_card_tests/test_autocudagraph.py:859]: TODO: 记录一个 Paddle 的 BUG，很离谱，这里如果不带 `.cpu()` 

```
            grads_eager.append(model_eager.head2.weight.grad.clone())
```

每次跑到固定位置，eager 的 grad 就回变成全 0！待排查具体原因

**Review by lshpku — APPROVED** [review]
